parser-typechecker/src/Unison/Codebase/Causal.hs

```diff
@@ -34,7 +30,7 @@ import           Util                           ( bind2 )
 * `cons : a -> Causal a -> Causal a`, satisfying `head (cons hd tl) == hd` and
           also `before tl (cons hd tl)`.
 * `merge : CommutativeSemigroup a => Causal a -> Causal a -> Causal a`, which is
-           associative and commutative and satisfies:
+           commutative (but not associative) and satisfies:
```

Comment:
Can we make it associative, passing a list of `Causal` to to `threeWayMerge` and `lca`?  (Doesn't have to be this PR.)

Reply:
We cannot, as the three-way merge operation is not associative and never was.

Reply:
I see; ok.

Reply:
Would you explain it for me though, and for posterity?  i.e. is there an n-way merge operation we could use that is associative?  I'm guessing no, but not sure why.

Reply:
The issue I was seeing was that merges were actually incorrect with the old algorithm. The reason is that when merging two n-way merge nodes, you don’t want to consider the LCA of all the nodes together. That might give you a very distant ancestor. I was seeing where merging two such nodes resulted in objects being “added back” after one branch removed them, since it appeared to the merge algorithm that they were new in one node and simply absent in the other.

Reply:
If you picked a single LCA for all the nodes being merged and then implemented an N-way LCA based merge then that would probably be associative. That’s not what you want though since the common LCA among all N nodes is worse than the LCA for the actual pairwise merges the user did. It’s also quadratic I think - every time a new thing gets merged into a chain of N merges you gotta repeat the work.

Fundamentally, LCA selection is dependent on grouping and affects results of a merge, so that kills associativity in general. Though in lots of specific instances, merge will be associative.

Reply:
Consider this case:

1. The root of the tree is `r`, an empty codebase.
2. The node `n` adds a definition `x`.
3. `n` is the LCA of `a` and `b`, where `a` adds `y` and `b` removes `x`.
4. `c` is a direct descendant of `r` and adds `z`.

Now: 
`a` has `{x,y}`
`b` has `{y}`.
`c` has `{z}`

A correct merge of `{a,c}` is `ac = {x,y,z}`.
A correct merge of `{b,c}`, is `bc = {y,z}`.
A correct merge of `{a,b}`, is `ab = {y}`.

Merge of `{a,bc}` is `{x,y,z}`
Merge of `{ab,c}` is `abbc = {y,z}`

Ergo, three-way merge is not associative.